modules/fundamental/src/vulnerability/service/mod.rs

```diff
@@ -155,6 +162,141 @@ impl VulnerabilityService {
+    {
+        if purls.is_empty() {
+            return Ok(Default::default());
+        }
+
+        let queries: Vec<Result<String, Error>> = purls
+            .iter()
+            .map(|p| {
+                let purl = match Purl::from_str(p) {
+                    Ok(purl) => purl,
+                    Err(e) => return Err(Error::Purl(e)),
+                };
+
+                let version = match purl.version.as_ref() {
+                    Some(v) => v,
+                    None => return Err(Error::Purl(PurlErr::MissingVersion(p.to_string()))),
+                };
+
+                let ns_condition = match purl.namespace.clone() {
+                    Some(namespace) => {
+                        let sql = "base_purl.namespace = $1";
+                        Statement::from_sql_and_values(
+                            connection.get_database_backend(),
+                            sql,
+                            [namespace.into()],
+                        )
+                    }
+                    None => Statement::from_string(
+                        connection.get_database_backend(),
+                        "base_purl.namespace IS NULL",
+                    ),
+                };
+
+                let name = purl.name;
+                let ty = purl.ty;
+                let values = [
+                    p.to_string().into(),
+                    name.into(),
+                    ty.into(),
+                    version.to_string().into(),
+                ];
+                let sql = format!(r#"
+                  SELECT $1 as requested_purl,
+                    vulnerability.id, vulnerability.title, vulnerability.reserved,
+                    vulnerability.published, vulnerability.modified, vulnerability.withdrawn, vulnerability.cwes
+                  FROM base_purl
+                    LEFT JOIN purl_status ON base_purl.id = purl_status.base_purl_id
+                    INNER JOIN version_range ON purl_status.version_range_id = version_range.id
+                    LEFT JOIN vulnerability ON purl_status.vulnerability_id = vulnerability.id
+                    INNER JOIN status ON purl_status.status_id = status.id
+                  WHERE {ns_condition}
+                    AND base_purl.name = $2
+                    AND base_purl.type = $3
+                    AND version_matches($4, version_range.*) = TRUE
+                    AND status.slug != 'fixed'"#);
+                let query =
+                    Statement::from_sql_and_values(connection.get_database_backend(), &sql, values);
+                Ok(query.to_string())
+            })
+            .collect();
+
+        let mut errors = Vec::new();
+
+        let query = queries
+            .into_iter()
+            .filter_map(|x| match x {
+                Ok(r) => Some(r),
+                Err(e) => {
+                    errors.push(e);
+                    None
+                }
+            })
+            .collect::<Vec<String>>()
```

Comment:
```suggestion
            .collect::<Result<Vec<_>,_>>()?
```
This is why I renamed `queries` to `query` above. `collect` is powerful. It can turn a vector of results into a result of vectors. And that `?` is what's returning the first error, if any. I'm using [this idiom](https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect) here.